File: 28_count_palindromic_substrings.py
"""
Task 28: Count all palindromic substrings in a string.
Uses expand-around-center O(n^2).
"""
import sys

def count_pal_subs(s: str) -> int:
    n = len(s); ans = 0
    for i in range(n):
        l=r=i
        while l>=0 and r<n and s[l]==s[r]: ans+=1; l-=1; r+=1
        l,i2 = i, i+1
        r = i2
        while l>=0 and r<n and s[l]==s[r]: ans+=1; l-=1; r+=1
    return ans

if __name__ == '__main__':
    s = sys.stdin.readline().rstrip('\n')
    print(count_pal_subs(s))

# Time Complexity: O(n^2)
# Space Complexity: O(1)

File: 28_count_palindromic_substrings.cpp
#include <bits/stdc++.h>
using namespace std;

/*
Task 28: Count all palindromic substrings in a string.
Uses expand-around-center O(n^2).
*/

int countPalindromes(const string& s)
{
    int n = s.size(), ans = 0;
    for (int i = 0; i < n; ++i)
    {
        int l = i, r = i;
        while (l >= 0 && r < n && s[l] == s[r])
        {
            ++ans;
            --l;
            ++r;
        }
        l = i;
        r = i + 1;
        while (l >= 0 && r < n && s[l] == s[r])
        {
            ++ans;
            --l;
            ++r;
        }
    }
    return ans;
}

int main()
{
    string s;
    if (!getline(cin, s))
        return 0;
    cout << countPalindromes(s) << '\n';
}

// Time Complexity: O(n^2)
// Space Complexity: O(1)
